subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileCompatibilityIntegrationTest.groovy

```diff
@@ -0,0 +1,587 @@
+        classJavaVersion(classFile("java", "custom", "Test.class")) == JavaVersion.toVersion(bytecodeOut)
+
+        where:
+        taskTarget | taskRelease | extensionTarget | taskSource | taskToolchain | sourceOut            | targetOut            | bytecodeOut
+        "9"        | "8"         | "11"            | "11"       | "11"          | "11"                 | "9"                  | "1.8"
+        "9"        | null        | "11"            | "8"        | "11"          | "8"                  | "9"                  | "9"
+        null       | "9"         | "11"            | "11"       | "11"          | "11"                 | "9"                  | "9"
+        null       | null        | "9"             | "8"        | "11"          | "8"                  | "9"                  | "9"
+        null       | null        | null            | "9"        | "11"          | "9"                  | "9"                  | "9"
+        null       | null        | null            | null       | "11"          | "11"                 | "11"                 | "11"
+        null       | null        | null            | null       | null          | currentJavaVersion() | currentJavaVersion() | currentJavaVersion()
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/22397")
+    def "uses source and target compatibility from toolchain defined by forkOptions #forkOption"() {
+        def currentJdk = Jvm.current()
+        def earlierJdk = AvailableJavaHomes.getDifferentVersion { it.languageVersion < currentJdk.javaVersion }
+        assumeNotNull(earlierJdk)
+
+        def path = TextUtil.normaliseFileSeparators(earlierJdk.javaHome.absolutePath.toString() + appendPath)
+
+        buildFile << """
+            apply plugin: "java"
+
+            compileJava {
+                options.fork = true
+                ${configure.replace("<path>", path)}
+            }
+
+            compileJava.doLast {
+                println "sourceCompatibility: '\${sourceCompatibility}'"
+                println "targetCompatibility: '\${targetCompatibility}'"
+            }
+        """
+
+        when:
+        withInstallations(earlierJdk).run(":compileJava", "--info")
+
+        then:
+        executedAndNotSkipped(":compileJava")
+        outputContains("Compiling with toolchain '${earlierJdk.javaHome.absolutePath}'")
+        outputContains("sourceCompatibility: '${earlierJdk.javaVersion}'")
+        outputContains("targetCompatibility: '${earlierJdk.javaVersion}'")
+        classJavaVersion(javaClassFile("Foo.class")) == earlierJdk.javaVersion
+
+        where:
+        forkOption   | configure                                       | appendPath
+        "java home"  | 'options.forkOptions.javaHome = file("<path>")' | ''
+        "executable" | 'options.forkOptions.executable = "<path>"'     | OperatingSystem.current().getExecutableName('/bin/javac')
+    }
+
+    def "uses matching compatibility options for source and target level"() {
```

Comment:
nitpick: I think this test name could be better.  I think what this is testing is that the source and target options match the toolchain, is that correct?  If so, then maybe something like "uses source and target levels that match the toolchain specified"?  Or whatever.  I just find the current name vague and indistinct - I had to read and think about the test to understand what was being tested.

Reply:
This test was there before my changes with this name and this implementation. I was also trying to understand it, but I don't think I succeeded.

As far as I can tell, it tests that setting toolchain to Java 11 allows compiling Java source code with language features from this Java version.

I already added other tests that cover this case, and wanted to delete this one, but I forgot. Thanks for catching it!